### PR TITLE
Byzcoinx 1186

### PR DESCRIPTION
**Pull request by ineiti**
This is the optimized implementation of ByzCoinX. It returns early whenever the threshold is reached. This works very well in a permissioned setting, however, in an unpermissioned setting, it can lead to censoring of slow nodes which will not get any mining reward.

It has been quite extensively tested in all test-suites, but there might still be some corner cases where it fails.

### CODE REVIEW

File: ftcosi/protocol/protocol.go [hunk 1]
```diff
@@ -1,14 +1,17 @@
 // Package protocol is the fault tolerant cosi protocol implementation.
 //
 // For more information on the protocol, please see
-// https://github.com/dedis/cothority/blob/master/ftcosi/protocol/README.md.
+// https:// github.com/dedis/cothority/blob/master/ftcosi/protocol/README.md.
```
Comment (kc1212) [inline ftcosi/protocol/protocol.go]: don't need the space.

File: byzcoinx/byzcoinx_test.go [hunk 2]
```diff
@@ -118,7 +118,8 @@ func TestBftCoSi(t *testing.T) {
 	require.Nil(t, err)
 
 	for _, n := range []int{1, 2, 4, 9, 20} {
-		runProtocol(t, n, 0, 0, protoName)
+		nbrFault := FaultThreshold(n)
+		runProtocol(t, n, nbrFault, 0, protoName)
```
Comment (kc1212) [inline byzcoinx/byzcoinx_test.go]: shouldn't this test check the scenario where we do not have any faults? because the faulty test is covered by `TestBftCoSiFault`

File: external/java/src/test/java/ch/epfl/dedis/lib/omniledger/contracts/EventLogInstanceTest.java [hunk 3]
```diff
@@ -70,11 +71,24 @@ void testLogMore() throws Exception {
...
-            assertEquals(event, event2);
+        boolean allOK = true;
+        for (int i = 0; i < 4; i++) {
+            allOK = true;
+            Thread.sleep(2 * ol.getConfig().getBlockInterval().toMillis());
+            for (InstanceId key : keys) {
+                try {
+                    Event event2 = el.get(key);
+                    assertEquals(event, event2);
+                } catch (CothorityCryptoException e){
```
Comment (kc1212) [inline external/java/src/test/java/ch/epfl/dedis/lib/omniledger/contracts/EventLogInstanceTest.java]: why don't we just let the exception be thrown? the test will fail and i think that achieves the same effect.

File: ftcosi/protocol/protocol.go [hunk 4]
```diff
@@ -127,72 +129,88 @@ func (p *FtCosi) Dispatch() error {
...
 	}
 
-	// signs the proposal
-	ok := <-verifyChan
-	if !ok {
-		// root should not fail the verification otherwise it would not have
-		// started the protocol
+	var secret kyber.Scalar
+	// verifies the proposal
+	verificationOk := <-verifyChan
```
Comment (kc1212) [inline ftcosi/protocol/protocol.go]: might be a good idea to add a timeout, just in case the verification function doesn't stop

File: ftcosi/protocol/protocol_test.go [hunk 5]
```diff
@@ -99,27 +100,24 @@ func TestProtocol(t *testing.T) {
 	}
 }
 
-// Tests unresponsive leaves in various tree configurations
-func TestUnresponsiveLeafs(t *testing.T) {
-	nodes := []int{3, 13, 24}
-	subtrees := []int{1, 2}
+// Tests various trees configurations
+func TestProtocolQuickAnswer(t *testing.T) {
```
Comment (kc1212) [inline ftcosi/protocol/protocol_test.go]: how do we know these are actually the quick answers, is it possible to check how long it takes for the signature to arrive?

File: external/java/src/test/java/ch/epfl/dedis/lib/omniledger/contracts/EventLogInstanceTest.java (hunk 3 above)
Comment (ineiti) [reply external/java/src/test/java/ch/epfl/dedis/lib/omniledger/contracts/EventLogInstanceTest.java]: Because we try four times to get all the proofs - so if we don't succeed in one of the first three rounds, we don't want the exception to be thrown.

File: ftcosi/protocol/protocol_test.go (hunk 5 above)
Comment (ineiti) [reply ftcosi/protocol/protocol_test.go]: I count the bits in the signature and make sure it's less than the total number of nodes - that should hold up most of the time, except with very few nodes and a very fast server.

File: ftcosi/protocol/protocol.go (hunk 4 above)
Comment (ineiti) [reply ftcosi/protocol/protocol.go]: added timeout while waiting for the verification.
